packages/suite/src/utils/suite/messageSystem.ts

```diff
@@ -0,0 +1,221 @@
+    const currentDate = Date.now();
+
+    const from = Date.parse(durationCondition.from);
+    const to = Date.parse(durationCondition.to);
+
+    return from <= currentDate && currentDate <= to;
+};
+
+export const validateVersionCompatibility = (
+    condition: Browser | OperatingSystem | Environment | Transport,
+    type: string | SuiteEnvironmentType,
+    version: string,
+): boolean => {
+    const conditionVersion = normalizeVersion(condition[type]);
+
+    if (conditionVersion === null) {
+        return false;
+    }
+
+    return semver.satisfies(version, conditionVersion);
+};
+
+export const validateSettingsCompatibility = (
+    settingsCondition: Settings[],
+    currentSettings: CurrentSettings,
+): boolean => {
+    const settings: {
+        [key: string]: any;
+    } = currentSettings.enabledNetworks.reduce((o, key) => Object.assign(o, { [key]: true }), {
+        tor: currentSettings.tor,
+    });
+
+    return settingsCondition.some(settingCondition =>
+        Object.entries(settingCondition).every(
+            ([key, value]: [string, boolean | unknown]) =>
+                settings[key] === value || (!value && settings[key] === undefined),
+        ),
+    );
+};
+
+export const validateTransportCompatibility = (
+    transportCondition: Transport,
+    transport?: Partial<TransportInfo>,
+): boolean => {
+    if (!transport || !transport.type || !transport.version) {
+        return false;
+    }
+
+    const { version } = transport;
+    const type = transport.type.toLowerCase();
+
+    return validateVersionCompatibility(transportCondition, type, version);
+};
+
+export const validateDeviceCompatibility = (
+    deviceConditions: Device[],
+    device?: TrezorDevice,
+): boolean => {
+    // if conditions are empty, then device should be empty
+    if (!deviceConditions.length) {
+        return !device;
+    }
+    if (!device || !device.features) {
+        return false;
+    }
+
+    const {
+        model,
+        vendor,
+        major_version: majorVersion,
+        minor_version: minorVersion,
+        patch_version: patchVersion,
+    } = device.features;
+
+    const deviceVersion = `${majorVersion}.${minorVersion}.${patchVersion}`;
+
+    return deviceConditions.some(
+        deviceCondition =>
+            deviceCondition.model.toLowerCase() === model.toLowerCase() &&
+            deviceCondition.vendor.toLowerCase() === vendor.toLowerCase() &&
+            semver.satisfies(deviceVersion, normalizeVersion(deviceCondition.firmware)!),
+    );
+};
+
+export const getValidMessages = (config: MessageSystem | null, options: Options): Message[] => {
+    if (!config) {
+        return [];
+    }
+
+    const { device, transport, settings } = options;
+
+    const ua = Bowser.getParser(getUserAgent());
+
+    const osDetail = ua.getOS();
+    const currentOsName = osDetail.name?.toLowerCase() || '';
+    const currentOsVersion = semver.valid(semver.coerce(osDetail.version)) || '';
```

Comment:
These in-house scripts are available in connect library https://github.com/trezor/connect/blob/develop/src/js/utils/versionUtils.js 

It is ready for comparison of two versions. However, in this case, I need to work with ranges, carets, tildes,...

Reply:
The in-house version validation is available in connect library https://github.com/trezor/connect/blob/develop/src/js/utils/versionUtils.js

However, it is not sufficient for this use case because it only allows comparing two versions. I need to compare ranges, work with tildes, carets,... 

I looked for alternatives but I did not find any more lightweight.